### PR TITLE
Globe occlusion for popups and improving marker performance on globe

### PR DESCRIPTION
**Pull request by SnailBones**
## Launch Checklist

https://github.com/mapbox/mapbox-gl-js/pull/11556 added support for marker and popup position on globe and marker occlusion behind globe. This PR adds occlusion for markers that are not attached to a popup.

Also improves the occlusion algorithm as described here: https://github.com/mapbox/mapbox-gl-js/pull/11556#pullrequestreview-925021709

Also fixes a minor bug where a marker behind terrain and in fog would have its fog opacity calculated at the point of the intersecting terrain, not the marker itself. i.e. in some cases marker opacity would increase when the marker was hidden behind terrain.

Other changes are refactoring for code readability and removing repetition.

A future optimization could include caching the result to reduce calls to `latLngToECEF`, though this may not be worth the added complexity.

 - [x] Globe occlusion for markeless popups.
 - [ ]  ~Fog support + terrain occlusion for markerless popups.~
 - [X] briefly describe the changes in this PR
 - [ ] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] manually test the debug page
 - [X] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [X] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Fix markers in fog sometimes becoming more visible when behind terrain. </changelog>`

### CODE REVIEW

File: src/geo/projection/globe_util.js [hunk 1]
```diff
@@ -439,24 +444,24 @@ function cameraPositionInECEF(tr: Transform): Array<number> {
...
 
-export function globeCenterToScreenPoint(tr: Transform): Point {
-    const pos = [0, 0, 0];
-    const matrix = mat4.identity(new Float64Array(16));
-    mat4.multiply(matrix, tr.pixelMatrix, tr.globeMatrix);
-    vec3.transformMat4(pos, pos, matrix);
-    return new Point(pos[0], pos[1]);
+export function isLngLatBehindGlobe(tr: Transform, lngLat: LngLat): boolean {
+    // We consider 1% past the horizon not occluded, this allows popups to be dragged around the globe edge without fading.
+    return (globeTiltAtLngLat(tr, lngLat) > Math.PI / 2 * 1.01);
```
Comment (karimnaaji) [inline src/geo/projection/globe_util.js:458]: 👍

File: src/ui/marker.js [hunk 2]
```diff
@@ -437,22 +435,17 @@ export default class Marker extends Evented {
...
+            opacity *= (1.0 - map._queryFogOpacity(mapLocation));
         }
 
-        const fogOpacity = map._queryFogOpacity(mapLocation);
-        opacity *= (1.0 - fogOpacity);
-        const pointerEvents = opacity ? 'auto' : 'none';
-
         this._element.style.opacity = `${opacity}`;
-        this._element.style.pointerEvents = pointerEvents;
+        this._element.style.pointerEvents = opacity ? 'auto' : 'none';
```
Comment (karimnaaji) [inline src/ui/marker.js:447]: nit: Prefer being explicit for this, improves readability and issues if someone changes opacity to be `undefined`.

```suggestion
        this._element.style.pointerEvents = opacity > 0 ? 'auto' : 'none';
```

File: src/ui/marker.js [hunk 3]
```diff
@@ -437,22 +435,17 @@ export default class Marker extends Evented {
...
-        if (map._usingGlobe()) {
-            opacity = this._occluded(mapLocation) ? 0 : 1;
-        } else if (map.transform._terrainEnabled() && map.getTerrain()) {
-            opacity = this._occluded(mapLocation) ? TERRAIN_OCCLUDED_OPACITY : 1;
+        let opacity = (map._usingGlobe() && isLngLatBehindGlobe(map.transform, this._lngLat)) ? 0 : 1;
+        if (opacity) {
+            if (map.transform._terrainEnabled() && map.getTerrain()) {
+                opacity = this._behindTerrain() ? TERRAIN_OCCLUDED_OPACITY : 1;
+            }
+            opacity *= (1.0 - map._queryFogOpacity(mapLocation));
```
Comment (karimnaaji) [inline src/ui/marker.js:443]: nit: Let's try to be more sequential and reduce overly branching.

```suggestion
        const behindGlobe = map._usingGlobe() && isLngLatBehindGlobe(map.transform, this._lngLat);
        const behindTerrain = map.transform._terrainEnabled() && map.getTerrain() && this._behindTerrain();

        const fogOcclusion = 1.0 - map._queryFogOpacity(mapLocation);

        const opacity = (behindGlobe ? 0 : (behindTerrain ? TERRAIN_OCCLUDED_OPACITY : 1)) * fogOcclusion;
```

**Review by karimnaaji — APPROVED** [review]
Looks good @SnailBones ! I've left a few nits, but the one missing thing before merging is adding a small test for popup occlusion, otherwise good to merge at your convenience.

File: src/ui/marker.js (hunk 3 above)
Comment (SnailBones) [reply src/ui/marker.js:443]: One concern with your rewrite is that checks for terrain and opacity will run even when the marker is occluded by the globe, shouldn't we skip those if possible?

It's possible to keep that optimization while rewriting these lines as a series of nested ternary operators, but IMO readability is worsened by doing this:

```js
        const behindGlobe = map._usingGlobe() && isLngLatBehindGlobe(map.transform, this._lngLat);
        const opacity = behindGlobe ? 0 :
            (map.transform._terrainEnabled() && map.getTerrain() && this._behindTerrain() ? TERRAIN_OCCLUDED_OPACITY : 1) *
            (1.0 - map._queryFogOpacity(mapLocation));
```

**Pull request merged**